File: patterns/human-readable-names/rules-python/human_readable_names_complex.py
import threading

from openhab import rule, Registry
from openhab.actions import Transformation
from openhab.triggers import GroupStateChangeTrigger

alert_timers = {}


@rule(triggers=[GroupStateChangeTrigger("gSensorStatus")])
class StatusAlert:
    def execute(self, module, input):
        event = input.get("event")
        if not event:
            return

        old_state = event.getOldItemState()
        if old_state is None or str(old_state) in ("NULL", "UNDEF"):
            return

        item_name = event.getItemName()
        if item_name in alert_timers:
            alert_timers[item_name].cancel()

        orig_state = str(Registry.getItem(item_name).getState())

        def on_expire():
            del alert_timers[item_name]
            if str(Registry.getItem(item_name).getState()) == orig_state:
                name = Transformation.transform("MAP", "admin.map", item_name) or item_name
                state_name = Transformation.transform("MAP", "admin.map", orig_state) or orig_state

                self.logger.info(name + " ist jetzt " + state_name)
                Registry.getItem(item_name + "_Alerted").postUpdate("ON")

        t = threading.Timer(60, on_expire)
        t.start()
        alert_timers[item_name] = t

File: patterns/human-readable-names/rules-js/human-readable-names-complex.js
const alertTimers = new Map();

rules.JSRule({
  name: "Status Alert",
  triggers: [triggers.GroupStateChangeTrigger("gSensorStatus")],
  execute: (event) => {
    if (event.oldState === undefined || event.oldState === null) return;

    const itemName = event.itemName;
    if (alertTimers.has(itemName)) alertTimers.get(itemName).cancel();

    const origState = String(items.getItem(itemName).state);

    const t = actions.ScriptExecution.createTimer(
      time.ZonedDateTime.now().plusMinutes(1),
      () => {
        alertTimers.delete(itemName);
        if (String(items.getItem(itemName).state) === origState) {
          const name = actions.Transformation.transform("MAP", "admin.map", itemName) || itemName;
          const stateName = actions.Transformation.transform("MAP", "admin.map", origState) || origState;

          console.log(name + " ist jetzt " + stateName);
          items.getItem(itemName + "_Alerted").postUpdate("ON");
        }
      }
    );
    alertTimers.set(itemName, t);
  }
});
